Cubical/Data/FinSet/Properties.agda

```diff
@@ -0,0 +1,151 @@
+
+private
+  variable
+    ℓ ℓ' : Level
+    A : Type ℓ
+
+-- infix operator to more conveniently compose equivalences
+
+_⋆_ = compEquiv
+
+infixr 30 _⋆_
+
+-- useful implications
+
+isFinSetFin : ∀ {n} → isFinSet (Fin n)
+isFinSetFin = ∣ _ , pathToEquiv refl ∣
+
+isFinSetUnit : isFinSet Unit
+isFinSetUnit = ∣ 1 , Unit≃Fin1 ∣
+
+isFinSet→Discrete : {X : Type ℓ} → isFinSet X → Discrete X
+isFinSet→Discrete p =
+  rec isPropDiscrete (λ (_ , p) → EquivPresDiscrete (invEquiv p) discreteFin) p
+
+isFinSet→isSet : {X : Type ℓ} → isFinSet X → isSet X
+isFinSet→isSet p = Discrete→isSet (isFinSet→Discrete p)
+
+isContr→isFinSet : {X : Type ℓ} → isContr X → isFinSet X
+isContr→isFinSet h = ∣ 1 , isContr→≃Unit* h ⋆ invEquiv (Unit≃Unit* ) ⋆ Unit≃Fin1 ∣
+
+isDecProp→isFinSet : {P : Type ℓ} → isProp P → Dec P → isFinSet P
+isDecProp→isFinSet h (yes p) = isContr→isFinSet (inhProp→isContr p h)
+isDecProp→isFinSet h (no ¬p) = ∣ 0 , uninhabEquiv ¬p ¬Fin0 ∣
+
+EquivPresFinSet : {X : Type ℓ}{Y : Type ℓ'} → X ≃ Y → isFinSet X → isFinSet Y
+EquivPresFinSet e p = rec isPropIsFinSet (λ (n , p) → ∣ n , compEquiv (invEquiv e) p ∣) p
+
+{-
+
+Alternative definition of finite sets
+
+A set is finite if it is merely equivalent to `Fin n` for some `n`. We
+can translate this to code in two ways: a truncated sigma of a nat and
+an equivalence, or a sigma of a nat and a truncated equivalence. We
+prove that both formulations are equivalent.
+
+-}
+
+isFinSet' : Type ℓ → Type ℓ
+isFinSet' A = Σ[ n ∈ ℕ ] ∥ A ≃ Fin n ∥
+
+FinSet' : (ℓ : Level) → Type (ℓ-suc ℓ)
+FinSet' ℓ = TypeWithStr _ isFinSet'
+
+isPropIsFinSet' : isProp (isFinSet' A)
+isPropIsFinSet' {A = A} (n , equivn) (m , equivm) =
+  Σ≡Prop (λ _ → isPropPropTrunc) n≡m
+  where
+    Fin-n≃Fin-m : ∥ Fin n ≃ Fin m ∥
+    Fin-n≃Fin-m = rec
+      isPropPropTrunc
+      (rec
+        (isPropΠ λ _ → isPropPropTrunc)
+        (λ hm hn → ∣ Fin n ≃⟨ invEquiv hn ⟩ A ≃⟨ hm ⟩ Fin m ■ ∣)
+        equivm
+      )
+      equivn
+
+    Fin-n≡Fin-m : ∥ Fin n ≡ Fin m ∥
+    Fin-n≡Fin-m = rec isPropPropTrunc (∣_∣ ∘ ua) Fin-n≃Fin-m
+
+    ∥n≡m∥ : ∥ n ≡ m ∥
+    ∥n≡m∥ = rec isPropPropTrunc (∣_∣ ∘ Fin-inj n m) Fin-n≡Fin-m
+
+    n≡m : n ≡ m
+    n≡m = rec (isSetℕ n m) (λ p → p) ∥n≡m∥
+
+isFinSet≡isFinSet' : isFinSet A ≡ isFinSet' A
+isFinSet≡isFinSet' {A = A} = hPropExt isPropIsFinSet isPropIsFinSet' to from
+  where
+    to : isFinSet A → isFinSet' A
+    to ∣ n , equiv ∣ = n , ∣ equiv ∣
+    to (squash p q i) = isPropIsFinSet' (to p) (to q) i
+
+    from : isFinSet' A → isFinSet A
+    from (n , ∣ isFinSet-A ∣) = ∣ n , isFinSet-A ∣
+    from (n , squash p q i) = isPropIsFinSet (from (n , p)) (from (n , q)) i
+
+FinSet≡FinSet' : FinSet ℓ ≡ FinSet' ℓ
+FinSet≡FinSet' = ua (isoToEquiv (iso to from to-from from-to))
+  where
+    to : FinSet ℓ → FinSet' ℓ
+    to (A , isFinSetA) = A , transport isFinSet≡isFinSet' isFinSetA
+
+    from : FinSet' ℓ → FinSet ℓ
+    from (A , isFinSet'A) = A , transport (sym isFinSet≡isFinSet') isFinSet'A
```

Comment:
Looking closer at `to`/`from` here makes me think that we should just put the two `to`/`from` in `isFinSet≡isFinSet'` on the top-level with good names and use them here. Seems silly to package the function up in an equivalence, use univalence, then transport, instead of just applying the functions